File: chapters/06_workspace_memory_and_search/v2_keyword_memory.py
from __future__ import annotations

from pathlib import Path


def load_memory_lines() -> list[str]:
    memory_file = Path(__file__).resolve().parent / "MEMORY.md"
    if not memory_file.exists():
        return []
    return [line.strip() for line in memory_file.read_text(encoding="utf-8").splitlines() if line.strip()]


def keyword_search(query: str, memory_lines: list[str], top_k: int = 2) -> list[str]:
    tokens = [token for token in query.lower().split() if token]
    scored: list[tuple[int, str]] = []

    for line in memory_lines:
        lowered = line.lower()
        score = sum(1 for token in tokens if token in lowered)
        if score > 0:
            scored.append((score, line))

    scored.sort(key=lambda item: item[0], reverse=True)
    return [line for _, line in scored[:top_k]]


def answer_with_keyword_memory(query: str) -> str:
    lines = load_memory_lines()
    hits = keyword_search(query, lines)
    if not hits:
        return "No memory result found."
    return f"Top memory: {hits[0]}"


def main() -> None:
    print(answer_with_keyword_memory("session key policy"))


if __name__ == "__main__":
    main()

File: chapters/06_workspace_memory_and_search/v2_keyword_memory.ts
// TypeScript mirror for Chapter 06: Workspace Memory and Search (v2)
// Run with: npx ts-node v2_keyword_memory.ts

import * as fs from "fs";
import * as path from "path";

function loadMemoryLines(): string[] {
  const memoryFile = path.join(__dirname, "MEMORY.md");
  if (!fs.existsSync(memoryFile)) {
    return [];
  }
  return fs
    .readFileSync(memoryFile, "utf-8")
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
}

function keywordSearch(query: string, memoryLines: string[], topK: number = 2): string[] {
  const tokens = query.toLowerCase().split(/\s+/).filter((t) => t);
  const scored: [number, string][] = [];

  for (const line of memoryLines) {
    const lowered = line.toLowerCase();
    const score = tokens.reduce((sum, token) => sum + (lowered.includes(token) ? 1 : 0), 0);
    if (score > 0) {
      scored.push([score, line]);
    }
  }

  scored.sort((a, b) => b[0] - a[0]);
  return scored.slice(0, topK).map(([, line]) => line);
}

function answerWithKeywordMemory(query: string): string {
  const lines = loadMemoryLines();
  const hits = keywordSearch(query, lines);
  if (hits.length === 0) {
    return "No memory result found.";
  }
  return `Top memory: ${hits[0]}`;
}

function main(): void {
  console.log(answerWithKeywordMemory("session key policy"));
}

main();
